File: app/src/couponConfig.js
// KUPON MERKEZİ sabitleri — tek yerden yönetilir.
// NOT: Bu bir bahis/ödeme sistemi DEĞİLDİR — kupon hazırlama/takip/başarı ölçme.
//
// FİYAT KURALI (kesin): Birim kolon bedeli KODA YAZILMAZ ve UYDURULMAZ.
// Bedel yalnız backend'in verdiği couponPricing {unitPrice, source, updatedAt}
// kaydından gelir (backend/data/coupon-pricing.json — kaynak+tarih zorunlu).
// Veri yoksa maliyet GÖSTERİLMEZ; "birim bedel verisi yok" denir.
export const COUPON_MAX_COLUMNS = 2500;      // resmi oyun kuralı: kolon üst sınırı
export const MAX_COUPONS_PER_WEEK = 10;      // haftalık kupon hakkı (uygulama kuralı)
export const LOCK_BEFORE_MS = 5 * 60 * 1000; // ilk maçtan 5 dk önce kilit

export const OUTCOMES = ['1', 'X', '2'];        // ekranda görünen
export const toOfficial = (o) => (o === 'X' ? '0' : o); // resmi eşleme: X → 0

// Kolon = seçilen işaret sayılarının çarpımı (tekli=1, çifte=2, üçlü=3).
export function columnCount(selections) {
  return (selections || []).reduce((n, s) => n * Math.max(1, (s.selectedOutcomes || []).length), 1);
}

// Maliyet: yalnız GERÇEK fiyat verisiyle hesaplanır; yoksa null (gösterilmez).
export function costOf(columns, pricing) {
  const u = Number(pricing?.unitPrice);
  return Number.isFinite(u) && u > 0 && Number.isFinite(columns) ? columns * u : null;
}

// Fiyat kaydı geçerli mi? (kaynak + tarih zorunlu — kaynağı belirsiz bedel kullanılmaz)
export function validPricing(p) {
  return !!(p && Number(p.unitPrice) > 0 && p.source && p.updatedAt);
}

// Bültenin kilit anı = ilk maçın başlangıcından 5 dk önce.
// NOT: Bu "bülten kilidi" artık yalnız DERECELİ kupon seçimi ve geçmiş hafta
// koruması için kullanılır. KULLANICI SEÇİMLERİ maç bazında kilitlenir
// (aşağıdaki matchLockAt/lockViolations) — kural: tercih, İLGİLİ MAÇ
// başlamadan önce kaydedilmiş olmalıdır. Böylece hafta ortasında bile
// başlamamış maçlara kupon kurulabilir; başlamış maça dokunulamaz.
export function lockAtOf(matches) {
  const times = (matches || []).map((m) => (m.date ? new Date(m.date).getTime() : NaN)).filter((t) => !Number.isNaN(t));
  return times.length ? Math.min(...times) - LOCK_BEFORE_MS : null;
}
export const isLockedNow = (lockAt) => (lockAt != null && Date.now() >= lockAt);

// ——— MAÇ BAZLI KİLİT ———
// Her maç KENDİ başlangıcından 5 dk önce kilitlenir.
export function matchLockAt(m) {
  const t = m?.date ? new Date(m.date).getTime() : NaN;
  return Number.isNaN(t) ? null : t - LOCK_BEFORE_MS;
}
export const isMatchLocked = (m, now = Date.now()) => {
  const la = matchLockAt(m);
  return la != null && now >= la;
};
/** { maçNo: kilitZamanıMs } — depo doğrulamasına verilir. */
export function lockMapOf(matches) {
  const map = {};
  for (const m of matches || []) {
    const la = matchLockAt(m);
    if (la != null && m.no != null) map[m.no] = la;
  }
  return map;
}
const _norm = (arr) => OUTCOMES.filter((o) => (arr || []).includes(o)).join('');
/**
 * DÜRÜSTLÜK DOĞRULAMASI: kilitlenmiş bir maçın seçimi, kilitten önceki
 * değerinden FARKLI OLAMAZ (yeni kuponda "önceki değer" boştur — yani
 * başlamış maça yeni seçim yapılamaz). Dönen: ihlal edilen maç no listesi.
 */
export function lockViolations({ selections, prevSelections = [], lockMap = {}, now = Date.now() }) {
  const prev = new Map((prevSelections || []).map((s) => [s.no, _norm(s.selectedOutcomes)]));
  const out = [];
  for (const s of selections || []) {
    const la = lockMap[s.no];
    if (la == null || now < la) continue;
    if (_norm(s.selectedOutcomes) !== (prev.get(s.no) || '')) out.push(s.no);
  }
  return out;
}

File: flutter/lib/core/coupon/coupon_config.dart
// KAYNAK: app/src/couponConfig.js — BİREBİR çeviri.
//
// KUPON MERKEZİ sabitleri — tek yerden yönetilir.
// NOT: Bu bir bahis/ödeme sistemi DEĞİLDİR — kupon hazırlama/takip/başarı ölçme.
//
// FİYAT KURALI (kesin): Birim kolon bedeli KODA YAZILMAZ ve UYDURULMAZ.
// Bedel yalnız backend'in verdiği couponPricing {unitPrice, source, updatedAt}
// kaydından gelir (backend/data/coupon-pricing.json — kaynak+tarih zorunlu).
// Veri yoksa maliyet GÖSTERİLMEZ; "birim bedel verisi yok" denir.

/// Resmi oyun kuralı: kolon üst sınırı.
const int kCouponMaxColumns = 2500;

/// Haftalık kupon hakkı (uygulama kuralı).
const int kMaxCouponsPerWeek = 10;

/// İlk maçtan 5 dk önce kilit.
const Duration kLockBefore = Duration(minutes: 5);

/// Ekranda görünen işaretler.
const List<String> kOutcomes = ['1', 'X', '2'];

/// Resmi eşleme: X → 0
String toOfficial(String o) => o == 'X' ? '0' : o;

/// Bir maçın kupon seçimi.
class CouponSelection {
  const CouponSelection({required this.no, this.selectedOutcomes = const []});

  final Object no;
  final List<String> selectedOutcomes;
}

/// Kolon = seçilen işaret sayılarının çarpımı (tekli=1, çifte=2, üçlü=3).
int columnCount(List<CouponSelection>? selections) {
  var n = 1;
  for (final s in selections ?? const <CouponSelection>[]) {
    final k = s.selectedOutcomes.length;
    n *= k < 1 ? 1 : k;
  }
  return n;
}

/// Maliyet: yalnız GERÇEK fiyat verisiyle hesaplanır; yoksa null
/// (gösterilmez — uydurma bedel yazılmaz).
double? costOf(int? columns, Map? pricing) {
  final u = pricing?['unitPrice'];
  final unit = u is num ? u.toDouble() : double.tryParse('$u');
  if (unit == null || !unit.isFinite || unit <= 0) return null;
  if (columns == null) return null;
  return columns * unit;
}

/// Fiyat kaydı geçerli mi? (kaynak + tarih zorunlu — kaynağı belirsiz bedel
/// kullanılmaz)
bool validPricing(Map? p) {
  if (p == null) return false;
  final u = p['unitPrice'];
  final unit = u is num ? u.toDouble() : double.tryParse('$u');
  return unit != null &&
      unit > 0 &&
      p['source'] != null &&
      '${p['source']}'.isNotEmpty &&
      p['updatedAt'] != null;
}

DateTime? _tarih(Object? iso) =>
    iso is String ? DateTime.tryParse(iso)?.toLocal() : null;

/// Bültenin kilit anı = ilk maçın başlangıcından 5 dk önce.
///
/// NOT: Bu "bülten kilidi" artık yalnız DERECELİ kupon seçimi ve geçmiş hafta
/// koruması için kullanılır. KULLANICI SEÇİMLERİ maç bazında kilitlenir
/// (aşağıdaki matchLockAt/lockViolations) — kural: tercih, İLGİLİ MAÇ
/// başlamadan önce kaydedilmiş olmalıdır. Böylece hafta ortasında bile
/// başlamamış maçlara kupon kurulabilir; başlamış maça dokunulamaz.
DateTime? lockAtOf(List? matches) {
  DateTime? enErken;
  for (final raw in matches ?? const []) {
    final t = _tarih((raw as Map)['date']);
    if (t == null) continue;
    if (enErken == null || t.isBefore(enErken)) enErken = t;
  }
  return enErken?.subtract(kLockBefore);
}

bool isLockedNow(DateTime? lockAt, {DateTime? now}) =>
    lockAt != null && !(now ?? DateTime.now()).isBefore(lockAt);

// ——— MAÇ BAZLI KİLİT ———
// Her maç KENDİ başlangıcından 5 dk önce kilitlenir.

DateTime? matchLockAt(Map? m) {
  final t = _tarih(m?['date']);
  return t?.subtract(kLockBefore);
}

bool isMatchLocked(Map? m, {DateTime? now}) {
  final la = matchLockAt(m);
  return la != null && !(now ?? DateTime.now()).isBefore(la);
}

/// `{ maçNo: kilitZamanı }` — depo doğrulamasına verilir.
Map<Object, DateTime> lockMapOf(List? matches) {
  final map = <Object, DateTime>{};
  for (final raw in matches ?? const []) {
    final m = raw as Map;
    final la = matchLockAt(m);
    if (la != null && m['no'] != null) map[m['no'] as Object] = la;
  }
  return map;
}

String _norm(List<String>? arr) =>
    kOutcomes.where((o) => (arr ?? const []).contains(o)).join();

/// DÜRÜSTLÜK DOĞRULAMASI: kilitlenmiş bir maçın seçimi, kilitten önceki
/// değerinden FARKLI OLAMAZ (yeni kuponda "önceki değer" boştur — yani
/// başlamış maça yeni seçim yapılamaz).
///
/// Dönen: ihlal edilen maç no listesi.
List<Object> lockViolations({
  required List<CouponSelection> selections,
  List<CouponSelection> prevSelections = const [],
  Map<Object, DateTime> lockMap = const {},
  DateTime? now,
}) {
  final simdi = now ?? DateTime.now();
  final prev = {
    for (final s in prevSelections) s.no: _norm(s.selectedOutcomes),
  };

  final out = <Object>[];
  for (final s in selections) {
    final la = lockMap[s.no];
    if (la == null || simdi.isBefore(la)) continue;
    if (_norm(s.selectedOutcomes) != (prev[s.no] ?? '')) out.add(s.no);
  }
  return out;
}
